includes/class-sensei-quiz.php

```diff
@@ -1280,17 +1280,45 @@ public static function start_quiz_questions_loop() {
 		global $sensei_question_loop;
 
 		// Initialise the questions loop object.
-		$sensei_question_loop['current']   = -1;
-		$sensei_question_loop['total']     = 0;
-		$sensei_question_loop['questions'] = array();
+		$sensei_question_loop['current']        = -1;
+		$sensei_question_loop['total']          = 0;
+		$sensei_question_loop['questions']      = [];
+		$sensei_question_loop['posts_per_page'] = -1;
+		$sensei_question_loop['current_page']   = 1;
+
+		// Get the pagination settings and populate the loop object.
+		$pagination = json_decode(
+			get_post_meta( get_the_ID(), '_pagination', true ),
+			true
+		);
 
-		$questions = Sensei()->lesson->lesson_quiz_questions( get_the_ID(), 'publish' );
+		if ( ! empty( $pagination['pagination_number'] ) ) {
+			$sensei_question_loop['posts_per_page'] = (int) $pagination['pagination_number'];
+		}
 
-		if ( count( $questions ) > 0 ) {
-			$sensei_question_loop['total']     = count( $questions );
-			$sensei_question_loop['questions'] = $questions;
-			$sensei_question_loop['quiz_id']   = get_the_ID();
+		// phpcs:ignore WordPress.Security.NonceVerification -- Argument is used for pagination in the frontend.
+		if ( ! empty( $_GET['questions-page'] ) ) {
+			// phpcs:ignore WordPress.Security.NonceVerification
+			$sensei_question_loop['current_page'] = max( 1, (int) $_GET['questions-page'] );
+		}
+
+		// Fetch the questions.
+		$all_questions = Sensei()->lesson->lesson_quiz_questions( get_the_ID(), 'publish' );
```

Comment:
This thing worries me a bit. Won't it be a performance problem for large quizzes?

Reply:
Yep, you are right. The issue is the `lesson_quiz_questions` method is making multiple queries, one for simple questions and n+1 for the question category type. In order to solve this, we need to change the architecture of how questions are stored in the database.  We've discussed this with @gkaragia and decided to just get all the questions for now. Any other solution will be too complex and will break backward compatibility. Maybe one sunny day (version 5.0?) we will clean this up and update the architecture. 🙂